learning-development/r.qmd

```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
+  
+  while (TRUE){
+    line <- readLines(con, n = 1)
+    
+    if ( length(line) == 0 ){
+      break
+    }
+    
+    line <- gsub("\\t", " ", line)
+    
+    if(grepl("--",line) == TRUE){
+      line <- paste(sub("--","/*",line),"*/")
+    }
+    
+    sql.string <- paste(sql.string, line)
+  }
+  
+  close(con)
+  return(sql.string)
+}
+
+# Execute SQL query and pull into R ====
+
+my_table <- dbGetQuery(con, getSQL("my_query_script.sql"))
+
+```
+
+If you're struggling to get your SQL scripts to execute from R, try adding the following two lines to the top of your SQL script to force the formatting required for it to work:
+
+
+```
+SET ANSI_PADDING OFF
+SET NOCOUNT ON;
+```
+
+---
+
+### Using dbplyr to create database queries
+
+If you're familiar with dplyr then using the dbplyr package to interact with a SQL database is a good way to go. It allows you to use common dplyr functions such as select(), filter(), mutate(), group_by(), summarise() and arrange() with a SQL database.
+
+As with sending a SQL script, you'll need to create the connection to the database first:
+
+```
+# Library calls ====
+
+library(odbc)
+library(DBI)
+
+# DB connection ====
+
+con <- DBI::dbConnect(odbc::odbc(),
+                      Driver = "ODBC Driver 17 for SQL Server",
+                      Server = "server_name",
+                      Database = "database_name",
+                      UID = "",
+                      PWD = "",
+                      Trusted_Connection = "Yes"
+)
+```
+
+Once you've established a connection, you can then create piped queries as you would with dplyr:
+
+```
+query_result <- collect(
+  tbl(con, Id(schema='dbo', table='table_name')) %>%
+  select(example_id_column, example_filter_column, example_indicator_column) %>%
+  filter(example_filter_column=='Filter entry 1', example_indicator_column <= 120)
+)
+```
+
+The key differences to just using dplyr are:
+
+- we use tbl() to point to the SQL table on the SQL database in place of supplying a data-frame
+- we use the collect() function to tell R to collate the commands and run them on the SQL database
+
+### Common issues interacting with a SQL database from within R
+
+#### DBI varchar(max) column ordering bug
+
+A known bug in DBI is that if a SQL table contains columns of type varchar(max), then any queries including those columns will fail if those columns aren't the last to be called in the query. This is a bit esoteric, but you can identify this issue if you're getting the following error message when you run your query:
+
+```
+Error: nanodbc/nanodbc.cpp:3166: 07009: [Microsoft][SQL Server Native Client 11.0]Invalid Descriptor Index 
+Warning message:
+In dbClearResult(rs) : Result already cleared
+```
+
+There are a couple of solutions to this:
+- avoid creating tables containing varchar(max) columns (i.e. always set a field length for your character fields)
+- make sure all varchar(max) columns are placed last in ay select query that you write
```

Comment:
Think you might need to add a blank line somewhere here (or at least that's my usual fudge - I should probably work out the actual reason / formatting to make it work!), as the bullet points aren't formatting correctly when you build it:
![image](https://github.com/dfe-analytical-services/analysts-guide/assets/52536248/f45f9697-a801-4615-86ce-1393fe246fe6)

Reply:
Done